src/main/java/com/microsoft/azure/relay/ClientWebSocket.java

```diff
@@ -323,6 +322,14 @@ public void onMessage(byte[] inputBytes, boolean isEnd) {
 	
 	@OnClose
 	public void onClose(Session session, CloseReason reason) {
+		CompletableFuture.runAsync(() -> {
+			try {
+				((LifeCycle) this.container).stop();	
```

Comment:
>((LifeCycle) this.container).stop(); [](start = 4, length = 36)

For some reason this must be put in a separate thread and not waited upon for it to work. If you ask me why, I'm not sure, but it does execute and do what it's supposed to do...

Reply:
does it deadlock if called on the thread invoking onClose? That wouldn't surprise me.

---
In reply to: [263108491](https://github.com/Azure/azure-relay-java/pull/29#discussion_r263108491) [](ancestors = 263108491)

Reply:
That's a likely cause. Are there better solutions than just simply putting it on a different thread?

---
In reply to: [263109139](https://github.com/Azure/azure-relay-java/pull/29#discussion_r263109139) [](ancestors = 263109139,263108491)

Reply:
This seems perfectly fine, or use executor.submit to use our HybridConnectionClient thread pool?

We have ActionItem.schedule but that's implemented with CF.runAsync anyhow!

---
In reply to: [263111543](https://github.com/Azure/azure-relay-java/pull/29#discussion_r263111543) [](ancestors = 263111543,263109139,263108491)

Reply:
Actually ActionItem should probably be removed. It's only used by InputQueue. That class should use its own executor instead.

---
In reply to: [263112694](https://github.com/Azure/azure-relay-java/pull/29#discussion_r263112694) [](ancestors = 263112694,263111543,263109139,263108491)